Use pytest in SQLLogic Python test runner

Flogex opened the pull request:
This PR adds pytest as test runner for the existing SQLLogic Python tests. _.test_, _.test\_slow_ and _.test\_coverage_ files in a given test directory are  are automatically discovered and added as test cases. The test runner also accepts a few options known from Catch2, like `--start-offset`, `--end-offset` or `--order`.

Usage:
- Default test run (excludes slow tests): `pytest 'tools/pythonpkg/tests/sqllogic/' --test-dir ./test`
- Running all tests: `pytest 'tools/pythonpkg/tests/sqllogic/' -m all --test-dir ./test`
- Running an individual test: `pytest 'tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/sql/order/test_limit_parameter.test]' --test-dir ./test`
- Running only tests with a particular marker, e.g. only slow tests: `pytest 'tools/pythonpkg/tests/sqllogic/' -m slow --capture=no --verbose --test-dir ./test`


Verbose output:
```
...
tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/extension/wrong_function_type.test] SKIPPED (Test is on SKIPPED_TESTS list)
tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/extension/test_tags.test] [19/3955] SKIPPED (/Users/me/mono/extensions/duckdb/te...)
tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/optimizer/zonemaps.test] [20/3955] PASSED
tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/optimizer/regex_optimizer.test] [21/3955] PASSED
tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/optimizer/constant_folding.test] [22/3955] PASSED
...
```

Non-verbose output:
```
tools/pythonpkg/tests/sqllogic/test_sqllogic.py sssFssssssssssssssss...........................s..............Fs...sF
```

This PR leaves the current test runner untouched and does not use the test runner yet in CI. For now, I simply copied over the code for the `SQLLogicTestExecutor` from the old test runner implementation and replaced the skip/fail methods with their pytest equivalents.

Flogex committed:
Use pytest in SQLLogic Python test runner

Integrate with pytest to be able to use a fully-fledged test runner.
Adds automatic discovery of .test files.
Adds a few command line options inspired by Catch2.

Tishj commented on tools/pythonpkg/tests/sqllogic/conftest.py:
```diff
@@ -0,0 +1,161 @@
+
+    # TODO: Add tests from extensions
+    # TODO: Do we have to handle --ignore and --ignore-glob and --deselect options?
+    test_script_extensions = [".test", ".test_slow", ".test_coverage"]
```
I don't think this should be looking for `test_slow` and `test_coverage` without explicit parameters indicating this
For example the `unittest` binary doesn't do this. To run *all* tests you would need to give it `"*"` as parameter.

Perhaps we can use the `-slow`, `-coverage`, `-fast` to toggle these. If none of these are set, `-fast` is assumed.

Flogex replied:
I have changed the test runner to exclude slow tests when no test cases are specified explicitly. That is, when running `pytest 'tools/pythonpkg/sqllogic/test_sqllogic.py'`, then the slow tests are not executed. But one can run them like:
- `pytest 'tools/pythonpkg/sqllogic/test_sqllogic.py' -m slow` or
- `pytest 'tools/pythonpkg/tests/sqllogic/test_sqllogic.py::test_sqllogic[test/sakila/test_sakila.test_slow]'`

https://github.com/duckdb/duckdb/pull/16685/commits/51178665ad7c62f998de090e433c107b4def80e8#diff-0b4cddc386386cfeb4cb1ee91dc3fa696f2eaffe31d3ec5d86987b6ea0200677R124-R146

pytest takes with the `-k` option a substring to find test cases that contain the substring. To run all tests including the slow ones, one option right now would be to run `pytest -k 'test/'` because "test/" is part of all test case names (of core DuckDB, ignoring extensions).
Do you think this is sufficient or would you prefer to have a `--run-all` option or something equivalent to Catch2's `"*"`?

Tishj replied:
Perhaps an explicit `--run-all` would be better then, `test/` feels a little wonky and also doesn't include all tests, if it ignores extensions

Flogex replied:
I added an "all" marker in https://github.com/duckdb/duckdb/pull/16685/commits/635bae32dcc7e2810f76ac3362bd6155b8aa52fa. This makes it possible to run all tests with `-m all`. Does this look good or do you prefer a command line option?

Tishj commented:
I think we *do* want to replace the existing method of calling the python sqllogictest with `pytest`, not have it live alongside it. To properly test this I think you want to start the nightlytest on your fork (as it won't run on PRs to the main remote)

Tishj commented on tools/pythonpkg/tests/sqllogic/test_sqllogic.py:
```diff
@@ -0,0 +1,131 @@
+        pytest.skip("Failed to parse SQLLogic script: " + str(e.message))
+        return
+
+    executor = SQLLogicTestExecutor()
```
This is missing the clearing of the temp directory
```suggestion
    executor = SQLLogicTestExecutor() # FIXME: where is the `args.build_dir` equivalent?
    if os.path.exists(TEST_DIRECTORY_PATH):
        shutil.rmtree(TEST_DIRECTORY_PATH)
```

Tishj replied:
Made a comment about build directory, this is needed for this (from `result.py`):
```py
    def get_extension_path(self, extension: str):
        if self.runner.build_directory is None:
            self.skiptest("Tried to load an extension, but --build-dir was not set!")
        root = self.runner.build_directory
        path = os.path.join(root, "extension", extension, f"{extension}.duckdb_extension")
        return path
```

Tishj commented on tools/pythonpkg/tests/sqllogic/conftest.py:
```diff
@@ -0,0 +1,161 @@
+import typing
+import warnings
+from .skipped_tests import SKIPPED_TESTS
+
```
As evidenced by the failing python CI run, this should not live inside `tools/pythonpkg/tests`, as that is already a `pytest` directory, it should probably live inside `tools/pythonpkg/sqllogic`

Flogex replied:
I have moved the files to tools/pythonpkg/sqllogic.

This resulted in an import error at first:
```
ImportError while importing test module '/Users/me/duckdb/tools/pythonpkg/sqllogic/test_sqllogic.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.13/3.13.2/Frameworks/Python.framework/Versions/3.13/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tools/pythonpkg/sqllogic/test_sqllogic.py:18: in <module>
    from sqllogictest.result import (
scripts/sqllogictest/result.py:33: in <module>
    from .logger import SQLLogicTestLogger
scripts/sqllogictest/logger.py:4: in <module>
    from duckdb import tokenize, token_type
tools/pythonpkg/duckdb/__init__.py:14: in <module>
    from .duckdb import (
E   ImportError: cannot import name 'SQLExpression' from 'duckdb.duckdb' (/Users/me/duckdb/tools/pythonpkg/duckdb/duckdb.cpython-313-darwin.so)
```

The reason is that tools/pythonpkg gets appended to `sys.path` when pytest starts so that pytest can then import the `sqllogic` package. (I think the sqllogic directory has to be a package to be able to import `skipped_tests.py` in `conftest.py`.)

I worked around this by changing the [import mode](https://docs.pytest.org/en/stable/explanation/pythonpath.html#import-modes) to `importlib` which doesn't change `sys.path`. Doing this in `conftest.py` is already to late because then the changes to the path have already happened. Therefore, I added a `pytest.ini` file which adds the `--import-mode=importlib` option when pytest is started.

Tishj replied:
Awesome, thanks for the investigation and the writeup, that will help with maintaining 👍

Flogex committed:
Exclude slow tests from default run, don't print progress if stdout is captured, & formatting

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Add --build-dir parameter and use pytest tmp_path as test directory

Flogex committed:
Move files to tools/pythonpkg/sqllogic

Flogex committed:
Formatting & remove tests from SKIPPED_TESTS again

Flogex committed:
Use new test runner in nightly tests

Flogex committed:
Fix directory path in nightly workflow

Flogex committed:
Don't set CI environment variable explicitly

GitHub Actions runners set this automatically

Flogex committed:
Make it possible to run all tests (including slow ones) with -m 'all'.

Flogex committed:
Set import-mode in pytest.ini

With import mode prepend, the tools/pythonpkg directory is added to the sys.path.
This leads to import errors when importing the duckdb package because there is
also a directory named tools/pythonpkg/duckdb.
Hence, setting the import mode to 'importlib' ('append' would also work).

Tishj commented:
Small changes from playing around with this:
```diff
diff --git a/scripts/sqllogictest/result.py b/scripts/sqllogictest/result.py
index e639f0baf8..1c74e82c97 100644
--- a/scripts/sqllogictest/result.py
+++ b/scripts/sqllogictest/result.py
@@ -99,7 +99,7 @@ class TestException(Exception):
     __slots__ = ['data', 'message', 'result']
 
     def __init__(self, data: SQLLogicStatementData, message: str, result: ExecuteResult):
-        self.message = f'{str(data)} {message}'
+        self.message = message
         super().__init__(self.message)
         self.data = data
         self.result = result
diff --git a/tools/pythonpkg/sqllogic/conftest.py b/tools/pythonpkg/sqllogic/conftest.py
index 498a020132..f0d6ab335a 100644
--- a/tools/pythonpkg/sqllogic/conftest.py
+++ b/tools/pythonpkg/sqllogic/conftest.py
@@ -165,7 +165,7 @@ def pytest_collection_modifyitems(session: pytest.Session, config: pytest.Config
     if start_offset < 0:
         raise ValueError("--start-offset must be a non-negative integer")
     elif end_offset < start_offset:
-        raise ValueError("--end-offset must be greater than or equal to --start-offset")
+        raise ValueError(f"--end-offset ({end_offset}) must be greater than or equal to --start-offset")
 
     max_end_offset = len(items) - 1
     if end_offset > max_end_offset:
```

The SQLStatementData no longer needs to be part of the error, the pytest name already contains this

Flogex commented:
Here is a run of the "Python SQLLogicTest Library" job in my fork: https://github.com/Flogex/duckdb/actions/runs/13967583581/job/39101440384?pr=7

Flogex committed:
Remove sqllogictest_python.py file

Flogex committed:
Add 'test' directory as default for --test-dir

Flogex commented on .github/workflows/NightlyTests.yml:
```diff
@@ -303,8 +303,10 @@ jobs:
+       env:
+         CI: 1
        run: |
-         python3 tools/pythonpkg/scripts/sqllogictest_python.py
```
Done. As far as I can see, the pytest runner covers almost all the functionality of the old one. The only thing that is missing is an equivalent to `--file-list` which accepts a file with a list of test cases to run.

- `--file-path` is replaced by passing the file path with `-k` or the name of the test case
- `--start-offset` and `--build-dir` are available
- Error reporting and skipping is handled by pytest
- `TEST_DIRECTORY_PATH` is handled by pytest (including cleaning up the old ones)

`SQLLogicTestExecutor` hasn't been changed much thus far and works as before.

Flogex committed:
Tishj's changes

Flogex committed:
Fix default for --test-dir

When using nargs=* and action=extend and a default value, then the default value
is always prepended to the list (of directory paths, in this case)

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Formatting

Flogex committed:
Remove outdated comment

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex commented:
@Tishj Here is the latest nightly test run: https://github.com/Flogex/duckdb/actions/runs/14022703354/job/39256726461

Three test failures
- test/sql/copy/return_stats_truncate.test
- test/sql/copy/return_stats.test
- test/sql/copy/parquet/writer/skip_empty_write.test
But those also fail in [main's nightly run](https://github.com/duckdb/duckdb/actions/runs/14024687113/job/39261371552), hence I think these are unrelated to this PR.

Tishj commented:
Yea agreed, I think the <REGEX> handling was changed in the cpp unittester and as always the python one wasn't changed, perhaps we can just skip them explicitly for now

Flogex committed:
Skip some copy tests in Python

Flogex committed:
Format skipped_tests.py

Flogex commented:
> Yea agreed, I think the `<REGEX>` handling was changed in the cpp unittester and as always the python one wasn't changed, perhaps we can just skip them explicitly for now

I excluded the three failing tests in https://github.com/duckdb/duckdb/pull/16685/commits/133ac9adc5f804ec7e9c91b133ebebe8c42346b4

Flogex committed:
Merge commit 'dc4bb03fd1ccde4575e2fb3091787af42afe9741' into sqllogic-pytes

Last nightlyt

Tishj reviewed (approved):
Thanks for working on this, it looks good to me now 👍

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Merge remote-tracking branch 'upstream/main' into sqllogic-pytest

Flogex committed:
Disable one more test in python runner
